pkg/controller/pipeline.go

```diff
@@ -0,0 +1,523 @@
+		),
+		capsule:        capsule,
+		currentObjects: map[ObjectKey]client.Object{},
+	}
+	if capsule.Status != nil {
+		p.generation = capsule.Status.ObservedGeneration
+	}
+	return p
+}
+
+func (p *Pipeline) Config() *configv1alpha1.OperatorConfig {
+	return p.config
+}
+
+func (p *Pipeline) Scheme() *runtime.Scheme {
+	return p.scheme
+}
+
+func (p *Pipeline) Capsule() *v1alpha2.Capsule {
+	return p.capsule
+}
+
+func (p *Pipeline) Client() client.Client {
+	return p.client
+}
+
+func (p *Pipeline) GetCurrent(key ObjectKey) client.Object {
+	o, ok := p.objects[key]
+	if !ok {
+		return nil
+	}
+
+	if o.Current == nil {
+		return nil
+	}
+
+	return o.Current.DeepCopyObject().(client.Object)
+}
+
+func (p *Pipeline) GetNew(key ObjectKey) client.Object {
+	o, ok := p.objects[key]
+	if !ok {
+		return nil
+	}
+
+	if o.New == nil {
+		return nil
+	}
+
+	return o.New.DeepCopyObject().(client.Object)
+}
+
+func (p *Pipeline) Set(key ObjectKey, obj client.Object) {
+	o, ok := p.objects[key]
+	if !ok {
+		o = &Object{}
+	}
+	o.New = obj
+	p.objects[key] = o
+}
+
+func (p *Pipeline) NamedObjectKey(name string, gvk schema.GroupVersionKind) ObjectKey {
+	return ObjectKey{
+		ObjectKey: types.NamespacedName{
+			Name:      name,
+			Namespace: p.capsule.Namespace,
+		},
+		GroupVersionKind: gvk,
+	}
+}
+
+func (p *Pipeline) ObjectKey(gvk schema.GroupVersionKind) ObjectKey {
+	return p.NamedObjectKey(p.capsule.Name, gvk)
+}
+
+func (p *Pipeline) MarkUsedResource(res v1alpha2.UsedResource) {
+	p.usedResources = append(p.usedResources, res)
+}
+
+func (p *Pipeline) AddStep(step Step) {
+	p.steps = append(p.steps, step)
+}
+
+func (p *Pipeline) Run(ctx context.Context) error {
+	// Read all status objects.
+	if s := p.capsule.Status; s != nil {
+		for _, r := range s.OwnedResources {
+			if r.Ref == nil {
+				continue
+			}
+			gk := schema.GroupKind{
+				Kind: r.Ref.Kind,
+			}
+			if r.Ref.APIGroup != nil {
+				gk.Group = *r.Ref.APIGroup
+			}
+
+			gvk, err := lookupGVK(gk)
+			if err != nil {
+				return err
+			}
+
+			o, err := p.scheme.New(gvk)
+			if err != nil {
+				return err
+			}
+
+			co, ok := o.(client.Object)
+			if !ok {
+				continue
+			}
+
+			co.SetName(r.Ref.Name)
+			co.SetNamespace(p.capsule.Namespace)
+			if err := p.client.Get(ctx, client.ObjectKeyFromObject(co), co); kerrors.IsNotFound(err) {
+				// Okay it doesn't exist, ignore the resource.
+				continue
+			} else if err != nil {
+				return err
+			}
+
+			p.currentObjects[p.NamedObjectKey(r.Ref.Name, gvk)] = co
+		}
+	}
+
+	for {
+		p.usedResources = nil
+		p.objects = map[ObjectKey]*Object{}
+		for k, o := range p.currentObjects {
+			p.objects[k] = &Object{
+				Current: o.DeepCopyObject().(client.Object),
+			}
+		}
+
+		p.logger.Info("run steps", "current_objects", maps.Keys(p.currentObjects))
+
+		for _, s := range p.steps {
+			if err := s.Apply(ctx, p); err != nil {
+				if err := p.updateStatus(ctx, nil, err); err != nil {
+					return err
+				}
```

Comment:
A couple times you use the pattern of calling `updateStatus` when we hit an error and then returning the original error.
We can wrap this entire loop in a function call, then call `updateStatus` afterwards if the function returns err. Then updating status on error is easier to maintain

Reply:
Done!